sympy/solvers/inequalities.py

```diff
@@ -982,3 +987,457 @@ def reduce_inequalities(inequalities, symbols=[]):
+    vectors satisfying the contraints is empty, then the problem is infeasible.
+
+    Example
+    =======
+
+    Suppose you want to maximize
+        x
+    subject to
+        x >= 10
+        x <= 9
+
+    Clearly it's not possible for x to satisfy the given contraints. If you try to use the linear_programing function on
+    a problem where it's not possible to satisfy the contraints, you will get this sort of exception.
+    """
+    pass
+
+
+def _pivot(M, i, j):
+    """
+    There are four rules for the pivot operation:
+    - The pivot quantity goes into its reciprocal
+    - Entries in the same row as the pivot are divided by the pivot
+    - Entries in the same column as the pivot are divided by the pivot and changed in sign.
+    - The remaining entries are reduced in value by the following: the product of the corresponding entries in the same
+    row and column as themselves and the pivot, divided by the pivot.
+
+    Example
+    =======
+
+    >>> from sympy.matrices.dense import Matrix
+    >>> from sympy.solvers.inequalities import _pivot
+    >>> m = Matrix([[3,1,5], \
+                    [2,-3,1], \
+                    [0,3,1]])
+    >>> _pivot(m, 1,0) # pivot around 2, which is in row 1 and col 0
+    Matrix([
+    [-3/2, 11/2, 7/2],
+    [ 1/2, -3/2, 1/2],
+    [   0,    3,   1]])
+    """
+    Mi, Mj, Mij = M[i,:], M[:,j], M[i,j]
+    if Mij == 0:
+        raise ZeroDivisionError("Tried to pivot about zero-valued entry.")
+    MM = M - Mj * (Mi / Mij)
+    MM[i,:] = Mi / Mij
+    MM[:,j] = -Mj / Mij
+    MM[i,j] = 1 / Mij
+    return MM
+
+
+def _simplex(M, skip_phase_2=False):
+    """
+    Simplex method with randomized pivoting
+
+    LINEAR PROGRAMMING: A Concise Introduction by Thomas S. Ferguson
+      - http://web.tecnico.ulisboa.pt/mcasquilho/acad/or/ftp/FergusonUCLA_LP.pdf
+
+    Explanation
+    ===========
+
+    When x is a column vector of variables, y is a column vector of dual variables,
+    and when the objective is either:
+
+    *) Maximizing Cx constrained to Ax <= B and x >= 0.
+    *) Minimizing y^{T}B constrained to y^{T}A >= C^{T} and y >= 0.
+
+    The method returns a triplet of solutions optimum, argmax and argmax_dual where
+    optimum is the optimum solution, and argmax/ argmax_dual give the
+    values for x and y, respectively.
+
+    Examples
+    ========
+
+    Suppose we want to maximize
+        3x+y
+    subject to
+        x+y <= 2
+
+    >>> from sympy.matrices.dense import Matrix
+    >>> from sympy.matrices.immutable import ImmutableMatrix
+    >>> from sympy.core.random import seed
+    >>> from sympy.solvers.inequalities import _simplex
+    >>> A = Matrix([[1,1]])
+    >>> B = Matrix([[2]])
+    >>> C = Matrix([[3, 1]])
+    >>> D = ImmutableMatrix([0])
+    >>> M = Matrix([[A, B], [-C, D]])
+    >>> M
+    Matrix([
+    [ 1,  1, 2],
+    [-3, -1, 0]])
+    >>> seed(47)
+    >>> optimum, argmax, argmax_dual = _simplex(M)
```

Comment:
Does the `seed` actually make a difference to the output?

Reply:
For these examples it shouldn't. I included it in docstrings just so people are aware they can use it.

If there are multiple argmax / argmax_dual, then it does matter.

Reply:
It would be better to be able to pass a seed argument to the function rather than calling a global seed like this.

Reply:
I don't think the seed should be served as public API at all.

This shouldnt really be using random at all.
There are other methods, like https://en.wikipedia.org/wiki/Bland%27s_rule which makes the reason that we don't need any random here, but have to be reserached by someone who's gonna implement this in production.

Reply:
From pg. 72-73 Gärtner, Bernd; Matoušek, Jiří (2006). Understanding and Using Linear Programming:
> Unfortunately, regarding efficiency, Bland’s rule is one of the slowest pivot rules and it is almost never used in practice. Another possibility can be found in the literature under the heading lexicographic rule, and here we only sketch it.
> [...]
> It can be shown that under the lexicographic rule, cycling is impossible. In very degenerate cases the lexicographic rule can be quite costly, since it may have to compute many components of the aforementioned vectors before the ties can eventually be broken.

Seems like the lexicographic rule may be the fastest rule that also avoids cycling (or at least that's what implied but not explicitly stated). However, the source that our simplex implementation is based off of recommends Bland's rule:

From pg. 34 of Thomas S. Ferguson, LINEAR PROGRAMMING: A Concise Introduction:
> In practical problems, the Modified Simplex Rule [the lexicographic rule] is never used. This is partly because
cycling in practical problems is rare, and so is not felt to be needed. But also there is a
very simple modification of the simplex method [Bland’s rule], due to Bland (“New finite pivoting rules
for the simplex method”, Math. of Oper. Res. 2, 1977, pp. 103-107), that avoids cycling,
and is yet very easy to include in a linear programming algorithm.

From what I can tell, [Z3](https://en.wikipedia.org/wiki/Z3_Theorem_Prover) implements the simplex method using Bland's rule. [This](https://theory.stanford.edu/~nikolaj/programmingz3.html#sec-solving-lra--linear-real-arithmetic) website says that Z3 uses this paper: [A Fast Linear-Arithmetic Solver for DPLL(T)](https://link.springer.com/chapter/10.1007/11817963_11).

From that paper:
> The main procedure of our algorithm is based on the dual Simplex and relies on Bland’s
pivot-selection rule to ensure termination.

I think despite its speed, Bland's rule is a good choice because of its simplicity and a big library like Z3 probably uses it.